Auto-summary for long chats, setting profiles, per chat settings, many UI tweaks.

Webifi opened the pull request:
Refactored a bunch of code while adding new features including:

- Auto-summarize to keep chats going before hitting the token limit 
- Setting profiles + export/import/clone  (Only three static starter profiles for now)
- Ability to use system prompt
- New settings to manage the above
- Clone chats
- Import/Export chats
- Settings saved per chat (plus profiles)
- To have a more robust selection, most icons were changed from emojis/unicode to font awesome.
- Consolidated some of the chat option (delete, clone, clear, plus new ones, into a 3-dot menu in the upper right)
- Many UI other fixes and enhancements.
- logit_bias can be manually added to setting profiles, and they work, but no UI for it yet.

For "profiles", while the ability the create a character for each by setting the system prompt is part of it, the main focus of the profiles are to have different setting profiles independent of the chat sessions.  So all settings are now stored and recalled with the chat session, and then also can be saved to custom "profiles", one of which can be selected as the default for new chat sessions. More/better example static profiles should be added. The can be exported, imported using the new 3-dot menu in the settings dialog, allowing profiles to be easily shared.

For deleting/editing past messages/prompts, currently a trash can will appear on hover.   For editing messages, just double click or double tap the the message.

For auto-summarization, there are new settings in the chat setting modal;  to enable it, the threshold for when it triggers, how many tokens it will try to summarize down to, how many prompts at the top and bottom should try to be preserved during the process, and the prompt used to generate the summary.  Once messages are summarized, they will be displayed slightly translucently, will no longer be editable or removeable. The completion they were boiled down to will be displayed below them.

Niek commented:
Thanks for the big PR! Lots of work in here, impressive!

I looked over it and have the following questions:
- The save chat/export chat are at bit confusing, wouldn't it be better to make that a single "save chat" item with a dialog or submenu asking to save as JSON or MD?
- There's some overlap between the pre-made prompts and the newly added profiles. I'm wondering if we can unify those (i.e. add all pre-mode prompts to the profiles dropdown).
- "Restart chat" and "clear chat messages" seem to do the same thing, with the exception that the former is also resetting the profile - I think this can be a bit confusing, probably better to get rid of one of these.
- Editing messages does not resubmit like the official ChatGPT UI does - this is probably confusing to users. What are your thoughts on this?

Good call on adding font-awesome icons btw, the emoji ones were getting too limited.

Webifi commented:
> * The save chat/export chat are at bit confusing, wouldn't it be better to make that a single "save chat" item with a dialog or submenu asking to save as JSON or MD?

Yeah, not exactly sure what to do here.  The MD export can't be re-imported, so it's not really a "save", so it needs to be viewed as a completely different thing. Unfortunately the current Bulma layout seem a bit restricted in the menu department, especially when tying to keep it mobile friendly.   I'll see what I can come up with help clarify and simplify this, but I really hate wrestling with Bulma.

> * There's some overlap between the pre-made prompts and the newly added profiles. I'm wondering if we can unify those (i.e. add all pre-mode prompts to the profiles dropdown).

I was worried about this, but the pre-made prompt DB you have are just prompts, the profiles are full settings profiles, so while it may seem similar on the surface, not a lot of actual overlap.  Perhaps a quick-select pulldown above the profile's system prompt could be used to allow inserting  some of the pre-made prompts into a profile?

> * "Restart chat" and "clear chat messages" seem to do the same thing, with the exception that the former is also resetting the profile - I think this can be a bit confusing, probably better to get rid of one of these.

"Restart chat" will clear messages and then re-inject the system prompt, and any trancing prompts, from the profile and auto-start (send it) if the profile is set to do so.  "Clear chat messages" will simply clear all messages.

> * Editing messages does not resubmit like the official ChatGPT UI does - this is probably confusing to users. What are your thoughts on this?

They will re-submit when you hit send -- more like the ChatGPT playground. (if you don't type anything in the input, the exiting prompts will simply be re-sent)  I think this is just a thing the user will need to get used to.  When you start to realize how it works, it's  much more powerful thing... however some type of truncate and send option on each user message could be handy.  UI layout/space limitations here kept me from adding it at this point.  Open to ideas. 

> Good call on adding font-awesome icons btw, the emoji ones were getting too limited.

Yeah, I debated between font-awesome and material design.  For svelte,  font-awesome seemed an easier path.

Fixed some things and added a couple new features.

Webifi commented:
For longer chat sessions, and with many saved chats, navigating the chat list on the left is problematic without hacking Bulma to death.

Any opposition to moving to a different UI library with a more traditional layout?  Like Svelte Material UI or others in that direction?

Webifi commented:
Just pushed a bunch of UI changes to hopefully clear up some things, yet not throw you too much. plus moved the settings modal to its own component.

> * The save chat/export chat are at bit confusing, wouldn't it be better to make that a single "save chat" item with a dialog or submenu asking to save as JSON or MD?

I've clarified the names in the upper right menu.  I've removed them all from the lower left.  Let me know if that helps.

I've also pinned the nav-bar and put the upper right menu in there.

> * There's some overlap between the pre-made prompts and the newly added profiles. I'm wondering if we can unify those (i.e. add all pre-mode prompts to the profiles dropdown).

Still thinking about this one.

> * "Restart chat" and "clear chat messages" seem to do the same thing, with the exception that the former is also resetting the profile - I think this can be a bit confusing, probably better to get rid of one of these.

Clarified the names a bit.  I think, depending on what's decided on the huge prompt DB, the clear messages option could be removed.

Webifi commented:
FYI, I'm testing some larger changes to the UI Here: (very beginning stages, need to add mobile support, a bunch of other things, only tested on chrome, but general idea is to get it a little closer to the official ChatGPT layout.)
https://github.com/Webifi/chatgpt-web/tree/UI-test
